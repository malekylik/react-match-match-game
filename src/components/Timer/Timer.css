.timer {
    width: 200px;
    padding-top: 10px;
    padding-bottom: 10px;
    font-size: 2.4rem;
    background-color: #ddd;
    text-align: center;
    border: 1px solid #bbb;
}
