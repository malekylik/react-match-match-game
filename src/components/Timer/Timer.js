import React, { Component } from 'react';
import PropTypes from 'prop-types';

import './Timer.css';

class Timer extends Component {
    timerId = null;

    start() {
        if (this.timerId === null) {
            this.timerId = setInterval(this.props.onTimeChange, 1000);
        }
    }

    terminate() {
        if (this.timerId !== null) {
            clearInterval(this.timerId);
            this.timerId = null;
        }
    }

    componentDidMount() {
        this.componentDidUpdate();
    }

    componentDidUpdate() {
        if (this.props.run) {
            this.start();
        } else {
            this.terminate();
        }
    }

    render() {
        const { time: { hours, minutes, seconds } } = this.props;

        return (
            <div className='timer'>
                {`${hours}:${minutes}:${seconds}`}
            </div>
        );
    }
}

Timer.defaultProps = {
    time: {
        hours: '00',
        minutes: '00',
        seconds: '00',
    }
};

Timer.propTypes = {
    time: PropTypes.shape({
        hours: PropTypes.string.isRequired,
        minutes: PropTypes.string.isRequired,
        seconds: PropTypes.string.isRequired,
    }).isRequired,
    run: PropTypes.bool.isRequired,
    onTimeChange: PropTypes.func,
};

export default Timer;
