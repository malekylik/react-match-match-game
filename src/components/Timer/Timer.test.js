import React from 'react';

import { shallow } from 'enzyme';

import Timer from './Timer';

describe('Timer', () => {
    const initialTime = {
        hours: String('00'),
        minutes: String('00'),
        seconds: String(10),
    };

    it('Should match snapshot', () => {
        expect(shallow(<Timer run={false} time={initialTime} />)).toMatchSnapshot();
    });

    describe('Timer.start()', () => {
        it("Should run timer", () => {
            const spy = jest.spyOn(Timer.prototype, 'start');
            const wrapper = shallow(<Timer run={true} time={initialTime} />);
            const timerId = wrapper.instance().timerId;
            
            expect(spy).toBeCalled();
            expect(timerId).not.toBeNull();

            spy.mockReset();
            spy.mockRestore();
        });

        it("Shouldn't run timer again if it is already running", () => {
            const wrapper = shallow(<Timer run={true} time={initialTime} />);
            const timerId = wrapper.instance().timerId;

            wrapper.setProps({ run: true });

            expect(wrapper.instance().timerId).toBe(timerId);
        });

        it("Should stop timer", () => {
            const spy = jest.spyOn(Timer.prototype, 'terminate');
            const wrapper = shallow(<Timer run={true} time={initialTime} />);

            wrapper.setProps({ run: false });
            
            expect(spy).toBeCalled();
            expect(wrapper.instance().timerId).toBeNull();

            spy.mockReset();
            spy.mockRestore();
        });
    });
});
