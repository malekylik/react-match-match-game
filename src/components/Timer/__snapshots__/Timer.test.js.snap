// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer Should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Timer
    run={false}
    time={
      Object {
        "hours": "00",
        "minutes": "00",
        "seconds": "10",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "00:00:10",
      "className": "timer",
    },
    "ref": null,
    "rendered": "00:00:10",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "00:00:10",
        "className": "timer",
      },
      "ref": null,
      "rendered": "00:00:10",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
